runelite-client/src/main/java/net/runelite/client/plugins/kourendlibrary/KourendLibraryOverlay.java

```diff
@@ -78,155 +76,170 @@ public Dimension render(Graphics2D g)
 		}
 
 		WorldPoint playerLoc = player.getWorldLocation();
-
-		if (playerLoc.distanceTo2D(LIBRARY_CENTER) > ROUGH_ENABLE_DISTANCE)
+		if (playerLoc.distanceTo2D(KourendLibraryPlugin.LIBRARY_CENTER) > KourendLibraryPlugin.ROUGH_ENABLE_DISTANCE)
 		{
 			return null;
 		}
 
-		List<Bookcase> allBookcases = library.getBookcasesOnLevel(client.getPlane());
+		List<Bookcase> allBookcases = library.getFloors().get(client.getPlane()).getBookcases();
 
-		if (allBookcases == null)
+		// Display a polygon under bookcases and the book(s) that might be or are present in them.
+		if (!allBookcases.isEmpty())
 		{
-			return null;
-		}
-
-		for (Bookcase bookcase : allBookcases)
-		{
-			// AABB
-			WorldPoint caseLoc = bookcase.getLocation();
-			if (Math.abs(playerLoc.getX() - caseLoc.getX()) > MAXIMUM_DISTANCE
-				|| Math.abs(playerLoc.getY() - caseLoc.getY()) > MAXIMUM_DISTANCE)
-			{
-				continue;
-			}
-
-			LocalPoint localBookcase = LocalPoint.fromWorld(client, caseLoc);
-			if (localBookcase == null)
-			{
-				continue;
-			}
-			Point screenBookcase = Perspective.worldToCanvas(client, localBookcase.getX(), localBookcase.getY(), caseLoc.getPlane(), 25);
-
-			if (screenBookcase != null)
-			{
-				boolean bookIsKnown = bookcase.isBookSet();
-				Book book = bookcase.getBook();
-				Set<Book> possible = bookcase.getPossibleBooks();
-				if (bookIsKnown && book == null)
+			allBookcases.stream()
```

Comment:
Try to not use streams when a simple iterator works in overlays, they have weird performance characteristics at times